File: jeong-hyuk/2월14일/N-Queen.py
def is_safe(board, row, col):

    for i in range(row):
        if board[i] == col or \
            board[i] - i == col - row or \
            board[i] + i == col + row:
            return False
    return True

def solve_n_queens(board, row, n, answer):
    if row == n:
        answer[0] += 1
        return

    for col in range(n):
        if is_safe(board, row, col):
            board[row] = col
            solve_n_queens(board, row + 1, n, answer)

def solution(n):
    answer = [0]
    board = [-1] * n
    solve_n_queens(board, 0, n, answer)
    return answer[0]

File: jja6312/2월14일/N-Queen.java
class Solution {
    int size;
    int cnt;

    public int solution(int n) {
        int answer = 0;

        // 깊이 우선 탐색을 써야겠다.
        size = n;
        cnt = 0;
        int[] board = new int[n];
        placeQueens(board, 0);
        answer = cnt;

        return answer;
    }

    private void placeQueens(int[] board, int row) {
        if (row == size) {
            cnt++;
        }
        for (int col = 0; col < size; col++) {
            if (isSafety(board, row, col)) {
                board[row] = col;
                placeQueens(board, row + 1);
            }
        }
    }

    private boolean isSafety(int[] board, int row, int col) {
        for (int i = 0; i < row; i++) {
            if (board[i] == col || Math.abs(row - i) == Math.abs(col - board[i])) {
                return false;
            }
        }
        return true;
    }
}